File: LeetCode/5. Longest Palindromic Substring/Solution.cpp
class Solution {
   public:
    string longestPalindrome(string s) {
        int n = s.size();
        if (n < 2) {
            return s;
        }
        vector<vector<bool>> dp(n, vector<bool>(n));
        int maxLen = 1;
        int maxStart = 0;
        for (int r = 1; r < n; r++) {
            for (int l = 0; l < r; l++) {
                if ((s[l] == s[r]) && (r - l <= 2 || dp[l + 1][r - 1])) {
                    dp[l][r] = true;
                    if (r - l + 1 > maxLen) {
                        maxLen = r - l + 1;
                        maxStart = l;
                    }
                }
            }
        }
        return s.substr(maxStart, maxLen);
    }
};

File: LeetCode/5. Longest Palindromic Substring/Solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n < 2:
            return s

        dp = [[False] * n for _ in range(n)]
        maxLen = 1
        maxStart = 0
        for r in range(1,n):
            for l in range(r):
                if (s[l]==s[r]) and (r-l<=2 or dp[l+1][r-1]):
                    dp[l][r]=True
                    if r-l+1 > maxLen:
                        maxLen = r-l+1
                        maxStart = l

        return s[maxStart:maxStart+maxLen]
